File: frontend/src/types/alerts.ts
/**
 * Strongly typed domain models for Real-Time Alerts & Threshold Warnings (Chunk M6-05).
 * Conforms strictly to:
 * - Chunk M3-11 Dynamic Red Zone & Threshold Trigger Engine (dynamic_contracts.py)
 * - Chunk M3-13 Alert database model (telemetry.py)
 * - Authoritative Regional Profile trigger thresholds (himalayan.py, coastal.py, riverine.py)
 */

export type DangerLevel =
  | "LOW"
  | "MODERATE"
  | "HIGH"
  | "VERY_HIGH"
  | "CRITICAL"
  | "UNINHABITABLE";

// ============================================================================
// Status, Indicator & Operator Enums (strictly mirroring M3-11)
// ============================================================================

export type DynamicTriggerStatus = "no_trigger" | "triggered" | "insufficient_data";

export type DynamicHazardIndicator =
  | "rainfall_24h"
  | "seismic_mmi"
  | "slope_deg"
  | "water_level_above_danger"
  | "landslide_debris_volume"
  | "landslide_activity"
  | "custom";

export type ComparisonOperator = ">" | ">=" | "<" | "<=" | "==";

export type AlertSeverity = "info" | "warning" | "severe" | "extreme";

export type AlertType =
  | "rainfall_threshold"
  | "landslide_warning"
  | "evacuation_notice"
  | "seismic_event"
  | "flood_breach"
  | "dynamic_red_zone";

// ============================================================================
// Explainability & Audit Trail Contracts (M3-11)
// ============================================================================

export interface SingleTriggerEvaluation {
  indicator: string;
  observed_value?: number | null;
  configured_threshold?: number | null;
  operator: ComparisonOperator | string;
  triggered: boolean;
  status: DynamicTriggerStatus;
  unit?: string | null;
  audit_note: string;
}

export interface DynamicRedZoneExplainability {
  decision_reason: string;
  summary: string;
  profile_id: string;
  profile_name?: string;
  trigger_evaluations: SingleTriggerEvaluation[];
  triggered_indicators: string[];
  missing_indicators: string[];
  source_observation_ids: string[];
  source_village_ids: string[];
  buffer_applied_m?: number | null;
  is_dissolved?: boolean;
  dissolved_count?: number;
  governance_notice: string;
}

// ============================================================================
// Operational Alert Model (Unified M3-11 Dynamic Candidate & M3-13 Alert)
// ============================================================================

export interface OperationalAlertItem {
  id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  danger_level?: DangerLevel | null;
  status: DynamicTriggerStatus;
  headline: string;
  message: string;
  village_id?: string | null;
  village_name?: string | null;
  district_id?: string | null;
  district_name?: string | null;
  coordinates?: [number, number] | null; // [longitude, latitude] WGS84
  indicator: DynamicHazardIndicator | string;
  observed_value?: number | null;
  configured_threshold?: number | null;
  operator: ComparisonOperator | string;
  unit?: string | null;
  buffer_m?: number | null;
  is_acknowledged: boolean;
  acknowledged_at?: string | null;
  acknowledged_by?: string | null;
  issued_at: string;
  expires_at?: string | null;
  candidate_id?: string | null;
  source_id?: string | null;
  is_synthetic: boolean;
  explainability?: DynamicRedZoneExplainability | null;
}

// ============================================================================
// Threshold Configuration Contract (M3-11 Profile Resolution)
// ============================================================================

export interface DynamicThresholdSummary {
  profile_id: string;
  profile_name: string;
  rainfall_trigger_24h_mm: number;
  seismic_trigger_mmi?: number | null;
  slope_trigger_min_deg: number;
  water_level_trigger_m_above_danger?: number | null;
  landslide_debris_volume_trigger_m3?: number | null;
  buffer_distance_m: number;
  default_danger_level: DangerLevel;
}

// ============================================================================
// Operational Filters & Overview Metrics
// ============================================================================

export interface AlertFilterCriteria {
  severity?: AlertSeverity | "all";
  status?: DynamicTriggerStatus | "all";
  indicator?: DynamicHazardIndicator | "all";
  is_acknowledged?: boolean | "all";
  search?: string;
}

export interface AlertSummaryMetrics {
  total_alerts: number;
  triggered_count: number;
  pending_acknowledgment: number;
  insufficient_data_count: number;
  critical_extreme_count: number;
}

File: backend/app/schemas/alerts.py
"""Pydantic validation and serialization schemas for real-time alerts (Chunk M3-11 / M6-05)."""

from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple
from pydantic import BaseModel, ConfigDict, Field


class AlertAcknowledgeRequest(BaseModel):
    """Payload for acknowledging an operational alert."""

    officer_name: str = Field("District Disaster Officer", description="Name of acknowledging officer")


class OperationalAlertRead(BaseModel):
    """Operational alert item conforming to M3-11 / M6-05 frontend contracts."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Unique alert identifier")
    alert_type: str = Field(..., description="Alert classification type")
    severity: str = Field(..., description="Severity level (info, warning, severe, extreme)")
    danger_level: Optional[str] = Field(None, description="Corroborated danger level")
    status: str = Field("triggered", description="Trigger status (no_trigger, triggered, insufficient_data)")
    headline: str = Field(..., description="Short broadcast headline")
    message: str = Field(..., description="Detailed emergency instruction")
    village_id: Optional[str] = Field(None, description="Associated village ID")
    village_name: Optional[str] = Field(None, description="Associated village name")
    district_id: Optional[str] = Field(None, description="Associated district ID")
    district_name: Optional[str] = Field(None, description="Associated district name")
    coordinates: Optional[Tuple[float, float]] = Field(None, description="[lon, lat] WGS84")
    indicator: str = Field("rainfall_24h", description="Physical trigger indicator")
    observed_value: Optional[float] = Field(None, description="Measured sensor/telemetry value")
    configured_threshold: Optional[float] = Field(None, description="Trigger limit threshold")
    operator: str = Field(">=", description="Comparison operator")
    unit: Optional[str] = Field("mm", description="Measurement unit")
    buffer_m: Optional[float] = Field(500.0, description="Hazard perimeter buffer in meters")
    is_acknowledged: bool = Field(False, description="Duty officer acknowledgment status")
    acknowledged_at: Optional[datetime] = Field(None, description="Acknowledgment timestamp")
    acknowledged_by: Optional[str] = Field(None, description="Officer who acknowledged the alert")
    issued_at: datetime = Field(..., description="Alert issuance timestamp")
    expires_at: Optional[datetime] = Field(None, description="Alert expiration timestamp")
    candidate_id: Optional[str] = Field(None, description="Associated dynamic red zone candidate ID")
    source_id: Optional[str] = Field(None, description="Telemetry provider source identifier")
    is_synthetic: bool = Field(True, description="Synthetic demo flag")
    explainability: Optional[Dict[str, Any]] = Field(None, description="Trigger audit and explainability details")
